mitodl/open-discussions
# Make jest faster

#### Pre-Flight checklist




#### What are the relevant tickets?
#4122 

#### What's this PR do?
This PR re-arranges our root `package.json` test commands a bit. But primarily it makes Jest faster by
- reducing the number of imports used by test files; a few new linting rules are added to help us avoid unnecessarily many imports in the future.
- also by setting `maxWorkers` a bit more intelligently on CI

Also, I updated Jest in the hope it would be faster. It wasn't. But might as well update.

See code comments for more details.

#### How should this be manually tested?
1. Tests should pass on CI. 
2. Tests should pass locally...
    - Run `yarn test` it should pass.
    - Run `yarn test-watch`, press "p" and start typing. You should get a live preview of matching test files. 
3. Test out the new linting rules: Add `import allTheFakers from "@faker-js/faker"` to any typescript file. You should get a linting error. 

#### Some numbers

- on CI:
    - **Previously:** Recent CI runs have been the 2-3 minute range: [145s](https://github.com/mitodl/open-discussions/actions/runs/5414841251/jobs/9842473735#step:11:206), [162s](https://github.com/mitodl/open-discussions/actions/runs/5414841341/jobs/9842473792#step:11:206), [221s](https://github.com/mitodl/open-discussions/actions/runs/5414841391/jobs/9842473811#step:11:206)
    - **Now:** Jest tests run in more like 60-70 seconds; e.g., [65s](https://github.com/mitodl/open-discussions/actions/runs/5416609894/jobs/9846561002#step:13:203) 
- locally, on my machine (after setting Docker to use new Virtios file syncing):
    -  **Previously:** Tests run 45-50 seconds, with very ~6 seconds of startup time for very simple test files in `infinite-corridor`
    - **Now:** Tests run in ~30 seconds, with ~2-3 seconds of startup time for very simple test files in `infinite-corridor`. 

At least for Macs, Docker is still a big slowdown: Tests run in about 12s on my host machine.

#### Some References
- Good insight on how Jest works and what causes it to be slow: https://blog.bitsrc.io/why-is-my-jest-suite-so-slow-2a4859bb9ac0
- Optimizing for CI: https://javascript.plainenglish.io/optimizing-jest-for-faster-ci-performance-with-github-actions-f4d7100c86c5
